Allow customizing debug color of Path3D.

ywmaa opened the pull request:
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->
Closes: https://github.com/godotengine/godot-proposals/issues/7853

ywmaa committed:
Path3D: custom debug color and custom debug show

ywmaa commented:
The solution is not working for some reason, so that's why I made as a draft.

YuriSizov commented:
Hey, I noticed that were making your first PRs recently. I recommend familiarizing with our [contribution guidelines](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md). Namely the preferred way to title commits and pull requests.

For example, for this feature the commit message could be "Allow to customize debug color of `Path3D`".

ywmaa commented:
Yeah, those are my first PRs.

> I recommend familiarizing with our [contribution guidelines](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md). Namely the preferred way to title commits and pull requests.
> 
> For example, for this feature the commit message could be "Allow to customize debug color of Path3D".

thanks that will help, I was using a message similar to the linux kernel stuff, lol.

smix8 commented:
Debug is split between Editor debug and runtime debug (because the Editor is not available at runtime).

The one that you see in the Editor and also in the screenshot of the proposal is from an Editor plugin.
`editor/plugins/path_2d_editor_plugin.cpp`
`editor/plugins/path_3d_editor_plugin.cpp`

The debug that you see at runtime with the "Visible paths" debug enabled is part of the Path2D and Path3D node.
`scene/2d/path_2d.cpp`
`scene/3d/path_3d.cpp`

A debug change should be added to both Path2D and Path3D as both nodes should keep feature parity wherever possible and reasonable.

ywmaa commented:
![image](https://github.com/godotengine/godot/assets/68971808/44fdb648-3d05-4bf4-8e1e-43d0539678e3)

@smix8 
you are right, turns out I was testing them wrong.
thanks, now I can continue, and probably make this PR ready.

ywmaa commented:
so while going through this, to change the color in the editor, it applies to all path3D, I think because the gizmos are just instances, is it worth to not use instances for the same material to have custom debug color  ?

smix8 commented:
That is why the e.g. NavigationAgent has a "debug_use_custom" bool.

When enabled it creates its own material copy so it does not affect all other instances that share the default material for performance.

ywmaa commented:
Just Rebased to make this PR ready, in case the team wants to merge it for 4.4

Calinou reviewed (changes requested):
Tested locally, there are some issues:

- The material used for custom colors should be unshaded, so that the shape appears at its intended color regardless of lighting conditions. Right now, the red path in the screenshot below uses `#ff0000` but appears quite dark due to this:

With sky | Black sky (no ambient light)
-|-
![image](https://github.com/user-attachments/assets/4c9fd4a8-4c65-4bbd-b30e-1de6fb317ac6) | ![image](https://github.com/user-attachments/assets/d285e215-ad32-4b84-8df1-f590bcdbb4a8)

- The material used for custom colors should be transparent. This yellow path should be invisible as its alpha component is 0, but it's visible:

![image](https://github.com/user-attachments/assets/7f9df24e-6252-4dc9-bf97-7f447bb6c319)

- Paths do not get brighter when selected (or darker when unselected) like the default path color does:

Default unselected | Default selected
-|-
![image](https://github.com/user-attachments/assets/fdc6bd07-cb39-4c2c-a01c-5225471008c6) | ![image](https://github.com/user-attachments/assets/eb7133c2-dcd0-4260-8918-18677785a9d4)

Custom unselected | Custom selected
-|-
![image](https://github.com/user-attachments/assets/45792903-533f-434a-8b7d-68e686287c78) | ![image](https://github.com/user-attachments/assets/c77293e6-5c1e-4efd-bdb5-60b3114a0d6b)

- The path color does not update immediately after you change it in the color picker (or even after pressing <kbd>Enter</kbd>). You need to do something else for the gizmo to update instead, but the update ought to happen as soon as possible.

- The editor view does not match the running project view, as **Show Path** being unchecked is ignored in the editor. Also, the main path is not colored when running the project; only the fishbones are:

Editor | Running project
-|-
![image](https://github.com/user-attachments/assets/bcd36242-5e86-4def-99bd-8a4503db995f) | ![image](https://github.com/user-attachments/assets/02584cf2-7c3a-430f-9282-a00d11933c2e)

- I'm not sure if there should be a **Show Path** option. Other nodes like RayCast2D/3D don't have it. If you want to hide a path, you can set its color to a fully transparent custom color, in which case the gizmo drawing should be skipped (to improve performance).

ywmaa commented:
is there a signal or a way to hook Color property realtime change to a function ?

ywmaa commented:
LOL, I think I did something wrong in the runtime material, the game crashes with the Path3D in the scene

ywmaa commented:
Now everything should work.

GustJc commented:
@ywmaa You should squash all your commits into a single one if you ever want this PR to be accepted.
You should probably take o look at the docs on how Godot likes to work (https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#modifying-a-pull-request)

> However, be aware that in our PR workflow, we favor commits that bring the codebase from one functional state to another functional state, without having intermediate commits fixing up bugs in your own code or style issues. Most of the time, we will prefer a single commit in a given PR

Godot likes to work with a single commit per PR, it helps keep track of things and revert features if something breaks.
When making modifications, you can just amend the changes and force push them to your branch.

Also, you can change the text of the PR to one of these words to auto-link the PR to the issue
https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

You actually almost got it, if you just change the "solves: #" to "Resolves godotengine/godot-proposals#7853" it will make a link to the original proposal on the github interface on the side.
This makes it easier to find things around, and it'll automatically close the proposal when this PR is accepted.

---
Would really love to see this feature merged someday.

ywmaa commented:
I totally forgot about this PR, I just squashed my commits.

Calinou reviewed (commented):
Tested locally, it works as expected. All feedback I pointed out previously has been addressed. Great work :slightly_smiling_face: 

Code looks good to me.

Calinou commented on editor/plugins/path_3d_editor_plugin.cpp:
```diff
@@ -284,6 +283,24 @@ void Path3DGizmo::redraw() {
+		debug_material->set_flag(StandardMaterial3D::FLAG_DISABLE_FOG, true);
+	}
+
+	// 0. Don't draw since nothing will be visible anyway.
+	if (path_color.a == 0.0) {
+		return;
+	}
```
The only thing I noticed is that if you set alpha to 0, the Path3D gizmo handles in the editor will disappear. I suggest removing this particular `return;` and only avoid creating the path, but not the handles further below.

ywmaa commented:
removed the early return, and just added a check before rendering the line and fishbones.

https://github.com/godotengine/godot/pull/82321/files#diff-75709a2df2a5a1969f7a3a2e533a1dd14c33751bd511f95c2e820724a2d1e890R303

akien-mga commented:
Thanks!
